lib/id3c/cli/command/etl/redcap_det.py

```diff
@@ -89,49 +93,50 @@ def decorated(*args, db: DatabaseSession, log_output: bool, **kwargs):
-
-                    # Only process REDCap record if all required instruments are complete
-                    incomplete_instruments = {
-                        instrument
-                            for instrument
-                            in required_instruments
-                            if not is_complete(instrument, redcap_record)
-                    }
-
-                    if incomplete_instruments:
-                        LOG.debug(f"The following required instruments «{incomplete_instruments}» are not yet marked complete. " + \
-                                  f"Skipping REDCap DET {det.id}")
-                        mark_skipped(db, det.id, etl_id)
-                        continue
-
-                    bundle = routine(det = det, redcap_record = redcap_record)
-
-                    if log_output:
-                        print(json.dumps(bundle, indent=2))
-
-                    if bundle:
-                        insert_fhir_bundle(db, bundle)
-                        mark_loaded(db, det.id, etl_id)
-                    else:
-                        mark_skipped(db, det.id, etl_id)
+            with pickled_cache(CACHE_FILE) as cache:
+                for det in redcap_det:
+                    with db.savepoint(f"redcap_det {det.id}"):
+                        LOG.info(f"Processing REDCap DET {det.id}")
+
+                        instrument = det.document['instrument']
+
+                        # Only pull REDCap record if the current instrument is complete
+                        if not is_complete(instrument, det.document):
+                            LOG.debug(f"Skipping incomplete or unverified REDCap DET {det.id}")
+                            mark_skipped(db, det.id, etl_id)
+                            continue
+
+                        redcap_record = get_redcap_record_from_det(det.document)
+
+                        if not redcap_record:
+                            LOG.debug(f"REDCap record is missing or invalid.  Skipping REDCap DET {det.id}")
+                            mark_skipped(db, det.id, etl_id)
+                            continue
+
+                        # Only process REDCap record if all required instruments are complete
+                        incomplete_instruments = {
+                            instrument
+                                for instrument
+                                in required_instruments
+                                if not is_complete(instrument, redcap_record)
+                        }
+
+                        if incomplete_instruments:
+                            LOG.debug(f"The following required instruments «{incomplete_instruments}» are not yet marked complete. " + \
+                                      f"Skipping REDCap DET {det.id}")
+                            mark_skipped(db, det.id, etl_id)
+                            continue
+
+                        bundle = routine(db = db, cache = cache, det = det, redcap_record = redcap_record)
```

Comment:
Similarly with the generic-ness of the `cache` parameter name here. I don't know if it's used for caching geocoding results, or REDCap records, or what. I don't think I'm convinced that the cache should be managed by this decorator instead of each routine which needs a cache.

Reply:
This does not need to block deploy, but should be addressed later.